### PR TITLE
Peers communication protocol change to WS - Closes #446 

### LINKED ISSUE #446: Rewrite peers module - phase two
Decoupling @4miners ideas from #406:

Phase 2 of 2: Change peers communication protocol from http to websockets.

### PR DESCRIPTION
**Pull request by MaciejBaj**
Change peer communication protocol from HTTP to web socket. 

Implementation is close connected to newly created library - [wamp-socket-cluster](https://github.com/LiskHQ/wamp-socket-cluster.git). Please review also changes made there.

Previous PR: https://github.com/LiskHQ/lisk/pull/580

Closes #446

### CODE REVIEW

**Comment by MaciejBaj** [comment]
Comments from previous (#580) PR:
```
app.js:268 - Load this information from config.json file
modules/system.js:127 - Get limit from helper/constants.js
test/api/peer/peer.transactions.main.ws.js: 32 - Does createTransaction() params address and password represent a key pair? if yes it should be randomAccount.password
```

File: Gruntfile.js [hunk 1]
```diff
@@ -41,6 +42,7 @@ module.exports = function (grunt) {
 						util.format('mkdir -p %s/logs', version_dir),
 						util.format('mkdir -p %s/pids', version_dir),
 						util.format('cp %s/app.js %s', release_dir, version_dir),
+						util.format('cp %s/workersController.js %s', release_dir, version_dir),
```
Comment (Isabello) [inline Gruntfile.js]:
```
util.format('cp %s/workersController.js %s', release_dir, version_dir),
```

should be

```
util.format('cp %s/workersController.js %s', __dirname, version_dir),
```

**Comment by Isabello** [comment]
peers unit tests are failing due to port expectation:

```

  peers
>> (node:20770) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): TypeError: Cannot read property 'add' of undefined
    update
      ��� should insert new peer
      ��� should insert new peer with only ip and port and state defined
      ��� should update existing peer
      ��� should not insert new peer if address changed but nonce is the same
      ��� should not insert new peer if address changed but nonce is the same
>> /home/lisk/jenkins/workspace/Lisk-Pipeline_PR-660-LSHH4C2LVF4LDLDW63Z24IWAEU5R66GBYPYQ3AU74UL4ALWPNNHQ/node_modules/js-nacl/lib/nacl_factory.js:27
>> var Module;if(!Module)Module=(typeof Module!=="undefined"?Module:null)||{};var moduleOverrides={};for(var key in Module){if(Module.hasOwnProperty(key)){moduleOverrides[key]=Module[key]}}var ENVIRONMENT_IS_WEB=false;var ENVIRONMENT_IS_WORKER=false;var ENVIRONMENT_IS_NODE=false;var ENVIRONMENT_IS_SHELL=false;if(Module["ENVIRONMENT"]){if(Module["ENVIRONMENT"]==="WEB"){ENVIRONMENT_IS_WEB=true}else if(Module["ENVIRONMENT"]==="WORKER"){ENVIRONMENT_IS_WORKER=true}else if(Module["ENVIRONMENT"]==="NODE"){ENVIRONMENT_IS_NODE=true}else if(Module["ENVIRONMENT"]==="SHELL"){ENVIRONMENT_IS_SHELL=true}else{throw new Error("The provided Module['ENVIRONMENT'] value is not valid. It must be one of: WEB|WORKER|NODE|SHELL.")}}else{ENVIRONMENT_IS_WEB=typeof window==="object";ENVIRONMENT_IS_WORKER=typeof importScripts==="function";ENVIRONMENT_IS_NODE=typeof process==="object"&&typeof r
>> AssertionError: expected '40.40.40.40:4001' to equal '40.40.40.40:4000'
>>     at Proxy.assertEqual (/home/lisk/jenkins/workspace/Lisk-Pipeline_PR-660-LSHH4C2LVF4LDLDW63Z24IWAEU5R66GBYPYQ3AU74UL4ALWPNNHQ/node_modules/chai/lib/chai/core/assertions.js:953:12)
>>     at Proxy.methodWrapper (/home/lisk/jenkins/workspace/Lisk-Pipeline_PR-660-LSHH4C2LVF4LDLDW63Z24IWAEU5R66GBYPYQ3AU74UL4ALWPNNHQ/node_modules/chai/lib/chai/utils/addMethod.js:57:25)
>>     at Proxy.<anonymous> (/home/lisk/jenkins/workspace/Lisk-Pipeline_PR-660-LSHH4C2LVF4LDLDW63Z24IWAEU5R66GBYPYQ3AU74UL4ALWPNNHQ/node_modules/chai-bignumber/chai-bignumber.js:40:20)
>>     at Proxy.overwritingMethodWrapper (/home/lisk/jenkins/workspace/Lisk-Pipeline_PR-660-LSHH4C2LVF4LDLDW63Z24IWAEU5R66GBYPYQ3AU74UL4ALWPNNHQ/node_modules/chai/lib/chai/utils/overwriteMethod.js:78:33)
>>     at /home/lisk/jenkins/workspace/Lisk-Pipeline_PR-660-LSHH4C2LVF4LDLDW63Z24IWAEU5R66GBYPYQ3AU74UL4ALWPNNHQ/test/unit/modules/peers.js:159:52
>>     at Immediate.<anonymous> (/home/lisk/jenkins/workspace/Lisk-Pipeline_PR-660-LSHH4C2LVF4LDLDW63Z24IWAEU5R66GBYPYQ3AU74UL4ALWPNNHQ/test/unit/modules/peers.js:30:11)
>>     at runCallback (timers.js:678:20)
>>     at tryOnImmediate (timers.js:645:5)
>>     at processImmediate [as _immediateCallback] (timers.js:617:5)
>> Exited with code: 7.
>> Error executing child process: Error: Process exited with code 7.
Warning: Task "exec:coverageSingle" failed. Use --force to continue.

```

File: test/config.json [hunk 2]
```diff
@@ -1,5 +1,6 @@
 {
     "port": 4000,
```
Comment (Isabello) [inline test/config.json]: 4001

File: test/config.json [hunk 3]
```diff
@@ -1,5 +1,6 @@
...
+    "httpPort": 5000,
```
Comment (Isabello) [inline test/config.json]: 4000

File: test/integration/configs/config.non-forge.json [hunk 4]
```diff
@@ -0,0 +1,99 @@
+{
+    "port": 4000,
```
Comment (Isabello) [inline test/integration/configs/config.non-forge.json]: 4001

File: test/integration/configs/config.non-forge.json [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    "httpPort": 5000,
```
Comment (Isabello) [inline test/integration/configs/config.non-forge.json]: 4000

File: config.json [hunk 6]
```diff
@@ -1,7 +1,8 @@
 {
     "port": 8000,
```
Comment (Isabello) [inline config.json]: 8001

File: config.json [hunk 7]
```diff
@@ -1,7 +1,8 @@
...
+    "httpPort": 8001,
```
Comment (Isabello) [inline config.json]: 8000

File: test/common/objectStubs.js [hunk 8]
```diff
@@ -8,8 +8,11 @@ var randomPeer = {
 	'ip': '40.40.40.40',
 	'os': 'unknown',
 	'port': 4000,
```
Comment (Isabello) [inline test/common/objectStubs.js]: switch these ports

File: test/common/wsCommunication.js [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+var wsCommunication = {
+
+	defaultConnectionState: null,
+
+	connect: function (ip, port, socketDefer, headers) {
+
+		var wsOptions = {
+			protocol: 'http',
+			hostname: ip || '127.0.0.1',
+			port: +port || 4000,
```
Comment (Isabello) [inline test/common/wsCommunication.js]: 4001

File: test/config.json [hunk 10]
```diff
@@ -47,7 +48,7 @@
         "enabled": true,
         "list": [{
             "ip": "127.0.0.1",
-            "port": 9999
+            "port": "4000"
```
Comment (Isabello) [inline test/config.json]: 4001

File: test/integration/configs/config.non-forge.json [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+                "delayAfter": 0,
+                "windowMs": 60000
+            }
+        }
+    },
+    "peers": {
+        "enabled": true,
+        "list": [{
+            "ip": "127.0.0.1",
+            "port": 4000
```
Comment (Isabello) [inline test/integration/configs/config.non-forge.json]: 4001

File: test/integration/configs/config.non-forge.json [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+        "options": {
+            "port": 443,
+            "address": "0.0.0.0",
+            "key": "./ssl/lisk.key",
+            "cert": "./ssl/lisk.crt"
+        }
+    },
+    "dapp": {
+        "masterrequired": true,
+        "masterpassword": "F6XK8Zn5EYr5",
```
Comment (Isabello) [inline test/integration/configs/config.non-forge.json]: remove

**Comment by Isabello** [comment]
Please ensure this log message shows ip instead of undefined:

```
2017-07-12 13:41:38 174: [dbg] 2017-07-12 13:41:38 | Received transaction 8623930823827538598 from peer undefined:4000
```

**Comment by Isabello** [comment]
Better bet is to rename `port` to `wsPort` so that its clear which port does what.

File: modules/peers.js [hunk 13]
```diff
@@ -407,22 +476,22 @@ Peers.prototype.list = function (options, cb) {
```
Comment (4miners) [inline modules/peers.js]: We still perform consensus calculations here in `Peers.prototype.list` - why if we have new functions for that?

File: modules/peers.js [hunk 14]
```diff
@@ -520,16 +589,28 @@ Peers.prototype.onPeersReady = function () {
 			},
 			updatePeers: function (seriesCb) {
 				var updated = 0;
-				var peers = library.logic.peers.list();
+				var peers = self.acceptable(library.logic.peers.list());
```
Comment (4miners) [inline modules/peers.js]: Why acceptable? We shouldn't have any peers in list that doesn't already satisfied acceptable.

File: modules/peers.js [hunk 15]
```diff
@@ -520,16 +589,28 @@ Peers.prototype.onPeersReady = function () {
...
 
 				library.logger.trace('Updating peers', {count: peers.length});
 
 				async.each(peers, function (peer, eachCb) {
 					// If peer is not banned and not been updated during last 3 sec - ping
 					if (peer && peer.state > 0 && (!peer.updated || Date.now() - peer.updated > 3000)) {
 						library.logger.trace('Updating peer', peer);
-						self.ping(peer, function (err) {
-							++updated;
+						peer.rpc.status(function (err, status) {
```
Comment (4miners) [inline modules/peers.js]: It repeats all the way - move to dedicated function or integrate with `peer.rpc.status`

File: modules/peers.js [hunk 16]
```diff
@@ -520,16 +589,28 @@ Peers.prototype.onPeersReady = function () {
...
+							if (err) {
+								peer.applyHeaders({state: Peer.STATE.DISCONNECTED});
+								library.logger.trace('Every 10sec peers check ping peer failed ' + peer.string, err);
```
Comment (4miners) [inline modules/peers.js]: Bad error description.

File: modules/peers.js [hunk 17]
```diff
@@ -545,7 +626,7 @@ Peers.prototype.onPeersReady = function () {
 		});
 	}
 	// Loop in 10sec intervals (5sec + 5sec connect timeout from pingPeer)
-	jobsQueue.register('peersDiscoveryAndUpdate', peersDiscoveryAndUpdate, 5000);
+	jobsQueue.register('peersDiscoveryAndUpdate', peersDiscoveryAndUpdate, 10000);
```
Comment (4miners) [inline modules/peers.js]: So now peers state is mainteaned in real time, so we don't need to recheck peers every 10 sec, can be moved to something like 30 or 60s.

File: app.js [hunk 18]
```diff
@@ -323,7 +331,7 @@ d.run(function () {
 			var queryParser = require('express-query-int');
 			var randomString = require('randomstring');
 
-			scope.nonce = randomString.generate(16);
+			scope.config.nonce = randomString.generate(16);
```
Comment (MaciejBaj) [inline app.js]: yes, it will be opened as separate PR

File: Jenkinsfile [hunk 19]
```diff
@@ -218,7 +218,7 @@ lock(resource: "Lisk-Core-Nodes", inversePrecedence: true) {
 			"Functional Peer - Blocks" : {
 				node('node-02'){
 					sh '''
-					export TEST=test/api/peer.blocks.js TEST_TYPE='FUNC'
+					export TEST=test/api/peer/peer.blocks.ws.js TEST_TYPE='FUNC'
```
Comment (MaciejBaj) [inline Jenkinsfile]: after discussion with @karmacoma  we decided to move WS tests to `transport` folder and rename them accordingly

File: api/common.js [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+module.exports.handshake = function (headers, cb) {
+
+	var peer = new Peer({
+		ip: headers.ip,
+		port: headers.port
+	});
+
+	headers = peer.applyHeaders(headers);
+
+	z_schema.validate(headers, schema.headers, function (error) {
```
Comment (MaciejBaj) [inline api/common.js]: thanks

File: api/ws/rpc/wsRPC.js [hunk 21]
```diff
@@ -0,0 +1,188 @@
+'use strict';
+
+var MasterWAMPServer = require('wamp-socket-cluster/MasterWAMPServer');
+var scClient = require('socketcluster-client');
+var WAMPClient = require('wamp-socket-cluster/WAMPClient');
+var Q = require('q');
```
Comment (MaciejBaj) [inline api/ws/rpc/wsRPC.js]: Bluebird not installed and I was using Q every time before. Not a big thing as es6 migration is still on the roadmap. Should I change it?

File: api/ws/rpc/wsRPC.js [hunk 22]
```diff
@@ -0,0 +1,188 @@
...
+
+	clientsConnectionsMap: {},
+	scClient: scClient,
+	wampClient: new WAMPClient(),
+	wsServer: null,
+
+	/**
+	 * @param {MasterWAMPServer} wsServer
+	 */
+	setServer: function (wsServer) {
```
Comment (MaciejBaj) [inline api/ws/rpc/wsRPC.js]: not really sure, it looks good with getter :)

File: api/ws/rpc/wsRPC.js [hunk 23]
```diff
@@ -0,0 +1,188 @@
...
+		port: connectionState.port,
+		protocol: 'http',
+		autoReconnect: true,
+		query: constants.getConst('headers')
+	};
+
+	var clientSocket = wsRPC.scClient.connect(options);
+	wsRPC.wampClient.upgradeToWAMP(clientSocket);
+
+
```
Comment (MaciejBaj) [inline api/ws/rpc/wsRPC.js]: thx

File: api/ws/transport.js [hunk 24]
```diff
@@ -0,0 +1,25 @@
+'use strict';
+
+var wsApi = require('../../helpers/wsApi');
+var wsRPC = require('./rpc/wsRPC').wsRPC;
+
+function TransportWSApi (transportModule, app, logger) {
+
+	wsRPC.getServer().registerRPCEndpoints({
+		acceptPeer: transportModule.internal.acceptPeer,
+		removePeer: transportModule.internal.removePeer,
```
Comment (MaciejBaj) [inline api/ws/transport.js]: In #654 branch I added divided acceptPeer action into insert and update, so it will be three then. Insert and removed accessible only through peer itself after receiving WS handshake, update after each new block application called by other peers. So they will b changed in corresponding PR created from #654 branch.

File: api/ws/transport.js [hunk 25]
```diff
@@ -0,0 +1,25 @@
...
+		ping: transportModule.internal.ping,
```
Comment (MaciejBaj) [inline api/ws/transport.js]: not, thanks

File: app.js [hunk 26]
```diff
@@ -281,6 +237,58 @@ d.run(function () {
...
 
+		webSocket: ['config', 'connect', 'logger', 'network', function (scope, cb) {
+			var webSocketConfig = {
+				workers: 1,
+				port: scope.config.port,
+				wsEngine: 'uws',
+				appName: 'lisk',
+				workerController: workersController.path,
+				perMessageDeflate: false,
+				secretKey: 'liskSecretKey',
```
Comment (MaciejBaj) [inline app.js]: just checked, it works with default secretKey setting as well so I will just remove it

File: helpers/constants.js [hunk 27]
```diff
@@ -91,5 +91,18 @@ module.exports = {
 	signatureLength: 196,
 	// WARNING: When changing totalAmount you also need to change getBlockRewards(int) SQL function!
 	totalAmount: 10000000000000000,
-	unconfirmedTransactionTimeOut: 10800 // 1080 blocks
+	unconfirmedTransactionTimeOut: 10800, // 1080 blocks
+
+	setConst: function (key, value) {
```
Comment (MaciejBaj) [inline helpers/constants.js]: getter is throwing errors when some property is unset, so the app will crash instead of working with undefined which will then could be a nightmare to track on

File: helpers/httpApi.js [hunk 28]
```diff
@@ -37,6 +37,7 @@ var middleware = {
 	errorLogger: function (logger, err, req, res, next) {
 		if (!err) { return next(); }
 		logger.error('API error ' + req.url, err.message);
+		console.trace(err);
```
Comment (MaciejBaj) [inline helpers/httpApi.js]: actually I left it by purpose to read our tests easier

File: helpers/wsApi.js [hunk 29]
```diff
@@ -0,0 +1,85 @@
...
+var constants = require('./constants');
+
+var z_schema = new Z_schema();
+
+var middleware = {
+
+	Handshake: function (system) {
+		return function (headers, cb) {
+
+			headers = headers || {};
```
Comment (MaciejBaj) [inline helpers/wsApi.js]: thanks

File: api/common.js [hunk 30]
```diff
@@ -0,0 +1,51 @@
+'use strict';
+
+var z_schema = require('../helpers/z_schema.js');
+
```
Comment (MaciejBaj) [inline api/common.js]: all file should be removed, sorry - logic is done in `helpers/wsApi.js`

File: helpers/wsApi.js [hunk 31]
```diff
@@ -0,0 +1,85 @@
...
+
+			var peer = new Peer(headers);
+			headers.state = Peer.STATE.CONNECTED;
+			headers = peer.applyHeaders(headers);
+
+			z_schema.validate(headers, schema.headers, function (error) {
```
Comment (MaciejBaj) [inline helpers/wsApi.js]: exactly, `api/common.js` should be removed. It is not used anywhere.

File: helpers/wsApi.js [hunk 32]
```diff
@@ -0,0 +1,85 @@
...
+	}
+};
+
+var extractHeaders = function (request) {
+	var headers = _.get(url.parse(request.url, true), 'query', {});
+	if (!headers) {
+		throw new Error('No headers specified');
+	}
+
+	headers.ip = request.remoteAddress.split(':').pop();
```
Comment (MaciejBaj) [inline helpers/wsApi.js]: the `remoteAddress` is in format: `::ffff:127.0.0.1`

File: logic/account.js [hunk 33]
```diff
@@ -669,119 +669,119 @@ Account.prototype.merge = function (address, diff, cb) {
 		if (diff[value] !== undefined) {
 			var trueValue = diff[value];
 			switch (self.conv[value]) {
-			case String:
```
Comment (MaciejBaj) [inline logic/account.js]: I added ` "SwitchCase": 1 ` indent rule

File: logic/broadcaster.js [hunk 34]
```diff
@@ -140,6 +138,11 @@ Broadcaster.prototype.enqueue = function (params, options) {
  * @return {setImmediateCallback} err | peers
  */
 Broadcaster.prototype.broadcast = function (params, options, cb) {
+	options.data.peer = library.logic.peers.me();
+	if (!options.data.peer) {
```
Comment (MaciejBaj) [inline logic/broadcaster.js]: good catch, the check is not needed anymore. Previously I was waiting for my own ip address being returned from a first peer connected.

File: logic/broadcaster.js [hunk 35]
```diff
@@ -151,17 +154,11 @@ Broadcaster.prototype.broadcast = function (params, options, cb) {
...
 		},
-		function getFromPeer (peers, waterCb) {
+		function sendToPeer (peers, waterCb) {
 			library.logger.debug('Begin broadcast', options);
-
-			if (params.limit === self.config.peerLimit) { 
-				peers = peers.slice(0, self.config.broadcastLimit);
-			}
-
+			if (params.limit === self.config.peerLimit) { peers = peers.slice(0, self.config.broadcastLimit); }
```
Comment (MaciejBaj) [inline logic/broadcaster.js]: cannot find any argument. I will correct it

File: api/ws/rpc/wsRPC.js (hunk 21 above)
Comment (4miners) [inline api/ws/rpc/wsRPC.js]: Yes, please - it will be consistent with changes in rounds rewrite branch.

File: api/ws/rpc/wsRPC.js (hunk 22 above)
Comment (4miners) [inline api/ws/rpc/wsRPC.js]: Yep, it looks good - but should be the reason, maybe move those exposed properties to private variables?

File: api/ws/transport.js (hunk 24 above)
Comment (4miners) [inline api/ws/transport.js]: ok, thanks for clarification :)

File: app.js (hunk 26 above)
Comment (4miners) [inline app.js]: But for what we need that secretKey, even default? Is used to encrypt connection, or what?

File: helpers/wsApi.js (hunk 32 above)
Comment (4miners) [inline helpers/wsApi.js]: So... can we support ipv6 then? :D

**Comment by Isabello** [comment]
As nice as ipv6 support is. We face potential network partitioning if we
have ipv4 only nodes...

File: app.js (hunk 26 above)
Comment (MaciejBaj) [inline app.js]:
```
A key which various SC processes will use to interact with scBroker broker processes securely, defaults to a 256 bits cryptographically random hex string
```

File: modules/peers.js [hunk 36]
```diff
@@ -42,10 +42,28 @@ function Peers (cb, scope) {
...
 			version: scope.config.version,
-		},
+			forging: {
+				force: scope.config.forging.force
+			}
+		}
 	};
 	self = this;
 
+	setInterval(function () {
```
Comment (MaciejBaj) [inline modules/peers.js]: yes

File: modules/loader.js [hunk 37]
```diff
@@ -729,13 +720,12 @@ Loader.prototype.getNetwork = function (cb) {
...
 		return setImmediate(cb, null, __private.network);
 	}
-
 	modules.peers.list({}, function (err, peers) {
 		if (err) {
 			return setImmediate(cb, err);
 		}
 
-		__private.network = __private.findGoodPeers(peers);
+		__private.network = Loader.prototype.findGoodPeers(peers);
```
Comment (MaciejBaj) [inline modules/loader.js]: thx

File: modules/loader.js [hunk 38]
```diff
@@ -656,26 +647,26 @@ __private.sync = function (cb) {
  * @private
  * @implements {modules.blocks.lastBlock.get}
  * @implements {library.logic.peers.create}
- * @param {number} heights
+ * @param {array<Peer>} peers
  * @return {Object} {height number, peers array}
  */
-__private.findGoodPeers = function (heights) {
+Loader.prototype.findGoodPeers = function (peers) {
```
Comment (MaciejBaj) [inline modules/loader.js]: i added test for this function

File: modules/peers.js (hunk 17 above)
Comment (MaciejBaj) [inline modules/peers.js]: @Isabello ?

File: modules/peers.js (hunk 15 above)
Comment (MaciejBaj) [inline modules/peers.js]: good point

File: modules/peers.js [hunk 39]
```diff
@@ -156,10 +185,25 @@ __private.insertSeeds = function (cb) {
...
+		peer.rpc.status(function (err, status) {
+			if (err) {
+				peer.applyHeaders({state: Peer.STATE.DISCONNECTED});
+				library.logger.trace('Ping peer failed: ' + peer.string, err);
+			} else {
+				peer.applyHeaders({
+					height: status.height,
+					broadhash: status.broadhash,
+					nonce: status.nonce,
+					state: Peer.STATE.CONNECTED //connected
```
Comment (MaciejBaj) [inline modules/peers.js]: after discovery phase there could be the situation that disconnected peer was drawn that is online now - we want to mark it as connected then

File: modules/system.js [hunk 40]
```diff
@@ -208,6 +224,13 @@ System.prototype.update = function (cb) {
 		getHeight: function (seriesCb) {
 			__private.height = modules.blocks.lastBlock.get().height;
 			return setImmediate(seriesCb);
+		},
+		setHeaders: function (seriesCb) {
+			constants.setConst('headers', _.assign(constants.headers || {}, {
```
Comment (MaciejBaj) [inline modules/system.js]: ok I agree. The better idea could be to pass getter for the headers from system module when it is necessary. Do you agree?

File: logic/peer.js [hunk 41]
```diff
@@ -42,12 +52,14 @@ Peer.prototype.properties = [
...
 	'clock',
 	'updated',
-	'nonce'
+	'nonce',
+	'httpPort'
 ];
 
 Peer.prototype.immutable = [
 	'ip',
 	'port',
```
Comment (MaciejBaj) [inline logic/peer.js]: it would be actually massive change as it used in many different places. Does anyone else agree?

File: helpers/wsApi.js (hunk 32 above)
Comment (MaciejBaj) [inline helpers/wsApi.js]: later? :)

File: helpers/peersManager.js [hunk 42]
```diff
@@ -0,0 +1,76 @@
...
+		delete this.nonceToAddressMap[peer.nonce];
+
+		this.addressToNonceMap[peer.string] = null;
+		delete this.addressToNonceMap[peer.string];
+
+		this.peers[peer.string] = null;
+		delete this.peers[peer.string];
+	};
+
+	this.update = function (peer) {
```
Comment (MaciejBaj) [inline helpers/peersManager.js]: why not, reusing the logic for checking if the peer is correct

File: test/integration/configs/config.non-forge.json (hunk 12 above)
Comment (MaciejBaj) [inline test/integration/configs/config.non-forge.json]: yes thanks

File: test/config.json (hunk 2 above)
Comment (MaciejBaj) [inline test/config.json]: 5000

File: test/common/objectStubs.js (hunk 8 above)
Comment (MaciejBaj) [inline test/common/objectStubs.js]: thanks

File: modules/peers.js [hunk 43]
```diff
@@ -256,6 +311,27 @@ __private.dbSave = function (cb) {
 	});
 };
 
+Peers.prototype.getConsensus = function (matched, active) {
+
+	if (library.config.forging.force) {
+		return undefined;
```
Comment (MaciejBaj) [inline modules/peers.js]: https://github.com/LiskHQ/lisk/blob/1.0.0/logic/broadcaster.js#L47

File: modules/peers.js [hunk 44]
```diff
@@ -256,6 +311,27 @@ __private.dbSave = function (cb) {
...
+	}
+
+	active = active || __private.getMatched({state: Peer.STATE.CONNECTED});
```
Comment (MaciejBaj) [inline modules/peers.js]: good point

File: modules/peers.js [hunk 45]
```diff
@@ -256,6 +311,27 @@ __private.dbSave = function (cb) {
...
+	matched = matched || __private.getMatched({broadhash: constants.headers.broadhash}, active);
+
+	active = active.slice(0, constants.maxPeers);
+	matched = matched.slice(0, constants.maxPeers);
+
+	var consensus = Math.round(matched.length / active.length * 100 * 1e2) / 1e2;
```
Comment (MaciejBaj) [inline modules/peers.js]: just copied it from previous calculation

File: modules/peers.js [hunk 46]
```diff
@@ -318,17 +372,33 @@ Peers.prototype.ping = function (peer, cb) {
 Peers.prototype.discover = function (cb) {
 	library.logger.trace('Peers->discover');
 	function getFromRandomPeer (waterCb) {
-		modules.transport.getFromRandomPeer({
-			api: '/list',
-			method: 'GET'
-		}, function (err, res) {
-			return setImmediate(waterCb, err, res);
+		self.list({limit: 1, allowedStates: [Peer.STATE.DISCONNECTED, Peer.STATE.CONNECTED]}, function (err, peers) {
```
Comment (MaciejBaj) [inline modules/peers.js]: yes

File: modules/peers.js (hunk 14 above)
Comment (MaciejBaj) [inline modules/peers.js]: yes, i will remove it

File: api/ws/rpc/wsRPC.js [hunk 47]
```diff
@@ -0,0 +1,188 @@
...
+				return setImmediate(cb, err);
+			});
+		};
+	};
+};
+
+module.exports = {
+	wsRPC: wsRPC,
+	ConnectionState: ConnectionState,
+	ClientRPCStub: ClientRPCStub
```
Comment (MaciejBaj) [inline api/ws/rpc/wsRPC.js]: Actually, this so I do not know if there is a reason to change it into a class with initialization.
The reason why wsRPC has a different structure than ClientRPCStub is that there can be only one WS server instance across the app, while many clients

File: modules/peers.js [hunk 48]
```diff
@@ -182,20 +226,31 @@ __private.dbLoad = function (cb) {
...
-					return setImmediate(eachCb);
+					return updatePeer(peer, eachCb);
 				}
-			} else {
-				self.ping(peer, function (err) {
-					++updated;
+				return setImmediate(eachCb);
+			}
+
+			return updatePeer(peer, eachCb);
```
Comment (MaciejBaj) [inline modules/peers.js]: I think so, after reading peers from DB we want to check connectivity with them

File: helpers/peersManager.js [hunk 49]
```diff
@@ -0,0 +1,76 @@
...
+
+	this.getByNonce = function (nonce) {
+		return this.peers[this.nonceToAddressMap[nonce]];
+	};
+
+	this.getNonce = function (address) {
+		return this.addressToNonceMap[address];
+	};
+
+	this.getAddress = function (nonce) {
```
Comment (MaciejBaj) [inline helpers/peersManager.js]: thanks, I will do it.

File: modules/peers.js [hunk 50]
```diff
@@ -107,7 +125,7 @@ __private.getByFilter = function (filter, cb) {
 	};
 
 	// Apply filters (by AND)
-	var peers = library.logic.peers.list(true);
+	var peers = library.logic.peers.list();
```
Comment (MaciejBaj) [inline modules/peers.js]: hmm, clear object wouldn't have my RPC property

File: modules/peers.js (hunk 17 above)
Comment (Isabello) [inline modules/peers.js]: I agree. We receive updates from peers regularly and therefore asking them for state may not be as neccessary as it was before. 60 seconds should be sufficient, and perhaps we could check only peers we havent heard from recently instead of all of our peers.

File: logic/peer.js (hunk 41 above)
Comment (Isabello) [inline logic/peer.js]: you know i do ;)

File: modules/peers.js [hunk 51]
```diff
@@ -156,10 +185,25 @@ __private.insertSeeds = function (cb) {
 	library.logger.trace('Peers->insertSeeds');
 	async.each(library.config.peers.list, function (peer, eachCb) {
 		peer = library.logic.peers.create(peer);
-		library.logger.trace('Processing seed peer: ' + peer.string);
-		self.ping(peer, function (err) {
-			++updated;
-			return setImmediate(eachCb);
+		library.logger.debug('Processing seed peer: ' + peer.string);
+		if (_.isEmpty(self.acceptable([peer]))) {
```
Comment (MaciejBaj) [inline modules/peers.js]: ok, I will change it

File: test/common/httpCommunication.js [hunk 52]
```diff
@@ -55,7 +55,7 @@ var httpCommunication = {
 
 			var request = node.popsicle.get({
 				url: node.baseUrl + '/peer/height',
-				headers: node.generatePeerHeaders(ip, 4000)
+				headers: node.generatePeerHeaders(ip, 5000)
```
Comment (Isabello) [inline test/common/httpCommunication.js]: Why hard code this? Can we not make it a variable?

File: api/ws/rpc/wsRPC.js (hunk 22 above)
Comment (MaciejBaj) [inline api/ws/rpc/wsRPC.js]: ok

File: config.json (hunk 6 above)
Comment (MaciejBaj) [inline config.json]: yes

File: config.json (hunk 7 above)
Comment (MaciejBaj) [inline config.json]: yes

File: modules/peers.js (hunk 15 above)
Comment (MaciejBaj) [inline modules/peers.js]: it is now part of function `__private.updatePeerStatus`

File: modules/peers.js [hunk 53]
```diff
@@ -265,49 +341,27 @@ __private.dbSave = function (cb) {
...
+	var frozenPeer = _.find(library.config.peers.list, function (__peer) {
+		return peer.ip === __peer.ip && peer.port === __peer.port;
 	});
 	if (frozenPeer) {
 		// FIXME: Keeping peer frozen is bad idea at all
-		library.logger.debug('Cannot remove frozen peer', pip + ':' + port);
-	} else {
-		return library.logic.peers.remove ({ip: pip, port: port});
+		library.logger.debug('Cannot remove frozen peer', peer.ip + ':' + peer.port);
+		peer.state = Peer.STATE.DISCONNECTED;
```
Comment (MaciejBaj) [inline modules/peers.js]: it is now part of function `__private.updatePeerStatus`

File: test/integration/configs/config.non-forge.json (hunk 11 above)
Comment (MaciejBaj) [inline test/integration/configs/config.non-forge.json]: thanks

File: app.js [hunk 54]
```diff
@@ -281,6 +237,58 @@ d.run(function () {
...
+				pingInterval: 5000,
+				// How many milliseconds to wait without receiving a ping
+				// before closing the socket
+				pingTimeout: 60000,
+				// Maximum amount of milliseconds to wait before force-killing
+				// a process after it was passed a 'SIGTERM' or 'SIGUSR2' signal
+				processTermTimeout: 10000,
```
Comment (MaciejBaj) [inline app.js]: yes, we are waiting `processTermTimeout` before calling `process.exit()` after receiving kill signal. New connections won't be accepted during that time.

**Comment by Isabello** [comment]
test is failing: see logs: https://jenkins.lisk.io/job/Lisk-Pipeline/job/PR-660/35/execution/node/240/log/

File: app.js (hunk 26 above)
Comment (4miners) [inline app.js]: Ok, thanks.

File: modules/blocks/process.js [hunk 55]
```diff
@@ -150,7 +147,7 @@ Process.prototype.loadBlocksOffset = function (limit, offset, verify, cb) {
 	var params = { limit: newLimit, offset: offset || 0 };
 
 	library.logger.debug('Loading blocks offset', {limit: limit, offset: offset, verify: verify});
-	// Execute in sequence via dbSequence
+	// Execute in sequence via dbSequence]
```
Comment (4miners) [inline modules/blocks/process.js]: ?

File: modules/blocks/process.js [hunk 56]
```diff
@@ -209,29 +206,22 @@ Process.prototype.loadBlocksOffset = function (limit, offset, verify, cb) {
  * @return {Object}   cb.lastValidBlock Normalized new last block
  */
 Process.prototype.loadBlocksFromPeer = function (peer, cb) {
-	// Set current last block as last valid block
```
Comment (4miners) [inline modules/blocks/process.js]: Why remove that and other comments below, they are no longer valid?

File: modules/peers.js (hunk 50 above)
Comment (4miners) [inline modules/peers.js]: Yep, but from that list API response is generated, isn't it? So API response will contain that RPC property and not contains properties that `library.logic.peers.list(true);` is normalizing. Please check that.

File: modules/blocks/process.js (hunk 55 above)
Comment (MaciejBaj) [inline modules/blocks/process.js]: thanks
